pyeudiw/storage/db_engine.py

```diff
@@ -83,10 +85,58 @@ def update_response_object(self, nonce: str, state: str, response_object: dict)
                     f"Cannot update document with nonce {nonce} and state {state} on {db_name}")
 
         if replica_count == 0:
-            raise Exception(
+            raise ReplicaError(
                 f"Cannot update document with state {state} and nonce {nonce} on any instance")
 
         return replica_count
+    
+    def find_chain(self, entity_id: str) -> Union[dict, None]:
+        for db_name, storage in self.storages:
+            try:
+                chain = storage.find_chain(entity_id)
+
+                if chain:
+                    return chain
+                
+            except Exception as e:
+                logger.critical(f"Error {str(e)}")
+                logger.critical(
+                    f"Cannot find chain {entity_id} on {db_name}")
+                
+        raise ReplicaError(f"Cannot find chain {entity_id} on any instance")
+    
+    def has_chain(self, entity_id: str):
+        if self.find_chain(entity_id) is not None:
+            return True
+        return False
+    
+    def add_chain(self, entity_id: str, trust_chain: list[str], exp: datetime) -> str:
+        replica_count = 0
+        for db_name, storage in self.storages:
+            try:
+                storage.add_chain(entity_id, trust_chain, exp)
+                replica_count += 1
+            except Exception as e:
+                logger.critical(
+                    "Cannot add chain with entity_id {entity_id} on database {db_name}")
+        
+        if replica_count == 0:
+            raise ReplicaError(
+                f"Cannot add chain {entity_id} on any instance")
+            
+    def update_chain(self, entity_id: str, trust_chain: list[str], exp: datetime) -> str:
+        replica_count = 0
+        for db_name, storage in self.storages:
+            try:
+                storage.update_chain(entity_id, trust_chain, exp)
+                replica_count += 1
+            except Exception as e:
+                logger.critical(
+                    "Cannot add chain with entity_id {entity_id} on database {db_name}")
+        
+        if replica_count == 0:
+            raise ReplicaError(
```

Comment:
this is not a Replica error since it doesn't take in account any replication mechanisms but just the missing trust attestation, please change accordingly this exception